drivers/dma/dma_xilinx_axi_dma.c

```diff
@@ -0,0 +1,1020 @@
+ * If is_first or is_last are NOT set, the buffer is considered part of a SG transfer consisting of
+ * multiple blocks. Otherwise, the block is one transfer.
+ */
+static inline int dma_xilinx_axi_dma_transfer_block(struct dma_xilinx_axi_dma_channel *channel,
+						    dma_addr_t buffer_addr, size_t block_size,
+						    bool is_first, bool is_last)
+{
+	volatile struct dma_xilinx_axi_dma_sg_descriptor *current_descriptor =
+		&channel->descriptors[channel->current_transfer_start_index];
+
+	dma_xilinx_axi_dma_disable_cache();
+
+#ifdef CONFIG_DMA_64BIT
+	current_descriptor->buffer_address = (uint32_t)buffer_addr & 0xffffffff;
+	current_descriptor->buffer_address_msb = (uint32_t)(buffer_addr >> 32);
+#else
+	current_descriptor->buffer_address = buffer_addr;
+#endif
+	current_descriptor->app0 = channel->sg_desc_app0;
+
+	if (block_size > UINT32_MAX) {
+		LOG_ERR("Too large block: %zu bytes!", block_size);
+		return -EINVAL;
+	}
+	/* clears the start of frame / end of frame flags as well */
+	current_descriptor->control = (uint32_t)block_size;
+
+	if (is_first) {
+		current_descriptor->control =
+			current_descriptor->control | XILINX_AXI_DMA_SG_DESCRIPTOR_CTRL_SOF_MASK;
+	}
+	if (is_last) {
+		current_descriptor->control =
+			current_descriptor->control | XILINX_AXI_DMA_SG_DESCRIPTOR_CTRL_EOF_MASK;
+	}
+
+	/* SG descriptor must be completed BEFORE hardware is made aware of it */
+	barrier_dmem_fence_full();
+
+	dma_xilinx_axi_dma_enable_cache();
+
+	return 0;
+}
+
+#ifdef CONFIG_DMA_64BIT
+static inline int dma_xilinx_axi_dma_config_reload(const struct device *dev, uint32_t channel,
+						   uint64_t src, uint64_t dst, size_t size)
+#else
+static inline int dma_xilinx_axi_dma_config_reload(const struct device *dev, uint32_t channel,
+						   uint32_t src, uint32_t dst, size_t size)
+#endif
+{
+	const struct dma_xilinx_axi_dma_config *cfg = dev->config;
+	struct dma_xilinx_axi_dma_data *data = dev->data;
+	struct dma_xilinx_axi_dma_channel *channel_data = &data->channels[channel];
+
+	if (channel >= cfg->channels) {
+		LOG_ERR("Invalid channel %" PRIu32 " - must be < %" PRIu32 "!", channel,
+			cfg->channels);
+		k_oops();
+		return -EINVAL;
+	}
+	/* one-block-at-a-time transfer */
+	return dma_xilinx_axi_dma_transfer_block(
+		channel_data, channel == XILINX_AXI_DMA_TX_CHANNEL_NUM ? src : dst, size, true,
+		true);
+}
+
+static int dma_xilinx_axi_dma_configure(const struct device *dev, uint32_t channel,
+					struct dma_config *dma_cfg)
+{
+	const struct dma_xilinx_axi_dma_config *cfg = dev->config;
+	struct dma_xilinx_axi_dma_data *data = dev->data;
+	struct dma_block_config *current_block = dma_cfg->head_block;
+	int ret = 0;
+	int block_count = 0;
+
+	struct dma_xilinx_axi_dma_register_space *regs =
+		(struct dma_xilinx_axi_dma_register_space *)cfg->reg;
+
+	if (channel >= cfg->channels) {
+		LOG_ERR("Invalid channel %" PRIu32 " - must be < %" PRIu32 "!", channel,
+			cfg->channels);
+		k_oops();
```

Comment:
configuration errors shouldn't be a k_oops(), the error should be propogated up the callstack to the application which can do what it likes, perhaps asserting or panicking (k_oops())

Reply:
I have removed the oops() and use appropriate error messages / return values instead.